src/pydvl/utils/parallel/map_reduce.py

```diff
@@ -158,6 +168,11 @@ def __init__(
         else:
             self.max_parallel_tasks = max_parallel_tasks
 
+        if isinstance(inputs, Sequence):
+            self.inputs_: Union[SequenceType[T], "ObjectRef[T]"] = inputs
+        else:
+            self.inputs_ = self.parallel_backend.put(inputs)
+
```

Comment:
I'm not sure I understand this logic: If the input is an ndarray, then we do _not_ `put()` it to the backend?

I know I keep going in circles around this question, but why can't we just `put` **all** inputs to the backend? If I'm not mireading, it makes for a cumbersome API, having to instantiate the parallel backend and `put()` the data to it before creating a parallel job.

Instead I'd expect `MapReduceJob` to know to `put()` the inputs and any arguments to the backend (except possibly primitive types, enums, strings, and a couple others)

Reply:
The issue with using `put()` on such objects that are meant to be split across jobs is that we would either have to call `get()` on them inside chunkify in order to split them or make `_chunkify()` a remote function just like the others and I think it would just make things harder to debug.

Reply:
Hmm... as it is, it is bit confusing as an interface. Either we put all arguments, or we don't, because this separate handling makes bogus usage rather likely and also hard to debug: if one doesn't put() objects they are automagically copied to every remote function, right? If one doesn't know this, things will be slow and it will be hard to know why.

What about calling `put` inside `chunkify` for each chunk separately? the chunks are computed in the main process, right? Then each process would only get the chunk it needs.

Reply:
The problem with your suggestion is that unless we keep track of the `ObjectRef` instances for each chunk across calls to `_chunkify()` then we won't be gaining anything from doing it that way.

Reply:
What we could do instead is call `put()` on the passed input regardless of what it is and wrap `_chunkify()` the same way we do for `map_func` and `reduce_func`.

This would also open the possibility to have a `chunk_func` argument that would allow the possibility to change the chunkification without having to create a new subclass.

Reply:
I'm not sure I understand the first objection, since chunkify is called by `map()` once. If you mean for repeated calls to map, then keeping track as you say should not be a problem at all, maybe even a simple `@lru_cache` can do the trick in a first approximation. If this sounds doable, we could try it, otherwise we can go for your suggestion and leave a comment for us to avoid copying all data to all workers.

As to the idea of custom chunkification, that sounds interesting, maybe we can create an issue for that too